File: artifacts/CS680/CS680_mod8_ENHANCED.py
import os
import time


class HashTableSearch:
    """
    Enhanced search engine using a Hash Table (Python dictionary)
    to provide O(1) average search complexity.
    """

    def __init__(self):
        self.lookup_table = {}
        self.record_count = 0

    def add_record(self, data, sequence):
        # We store the data string as the key.
        # Since the requirement is to search for a word and display it,
        # and exact match is implied, the hash table is ideal.
        # If multiple records have the same word, we'll store the first one encountered,
        # consistent with the original heap DFS behavior.
        if data not in self.lookup_table:
            self.lookup_table[data] = sequence
            self.record_count += 1

    def search(self, word):
        # Time the lookup to demonstrate performance gain.
        start_time = time.perf_counter()
        # In a hash table, the 'search' is an O(1) lookup.
        # We return 1 for 'records searched' to indicate the direct access.
        found_sequence = self.lookup_table.get(word)
        end_time = time.perf_counter()

        search_duration = end_time - start_time
        return found_sequence, 1, search_duration


def load_data(file_path):
    engine = HashTableSearch()
    try:
        if not os.path.exists(file_path):
            return None

        with open(file_path, "r", encoding="utf-8") as f:
            lines = f.readlines()
            if not lines:
                return engine

            # Skip header line (e.g., sequence|data)
            for line in lines[1:]:
                line = line.strip()
                if not line:
                    continue
                parts = line.split("|")
                if len(parts) == 2:
                    try:
                        seq = int(parts[0])
                        data = parts[1]
                        engine.add_record(data, seq)
                    except ValueError:
                        # Skip malformed lines
                        continue
        return engine
    except Exception as e:
        print(f"Error loading file: {e}")
        return None


def main():
    # File path configuration matching the original structure
    base_path = "Module Eight Activity Data Sets/Module 8 Data Set-"

    print("--- CS680 Module Eight Search (ENHANCED) ---")
    print("Architectural Optimization: Hash Table (O(1) Search)")

    dataset_num = input(
        "Enter the numbered data set you want to search (1 - 5): "
    ).strip()
    file_path = base_path + dataset_num + ".csv"

    if not os.path.exists(file_path):
        print(f"Error: File '{file_path}' not found.")
        return

    print(f"Loading data from {file_path} into an optimized Hash Table...")
    start_load = time.perf_counter()
    engine = load_data(file_path)
    end_load = time.perf_counter()

    if engine is None:
        return

    print(f"Data loaded successfully in {end_load - start_load:.4f} seconds.")
    print(f"Total unique records indexed: {engine.record_count}")

    while True:
        search_word = input(
            "\nEnter a word to search for (or 'quit' to exit): "
        ).strip()

        if search_word.lower() == "quit":
            print("Exiting search application.")
            break

        if not search_word:
            continue

        # Execute the optimized search
        found_seq, records_searched, duration = engine.search(search_word)

        print("\n--- Search Results ---")
        if found_seq is not None:
            # Result found. Labeled data field shown; integer sequence hidden.
            print(f"Data: {search_word}")
        else:
            print(f"The value '{search_word}' was not located within the data set.")

        print(f"Records searched: {records_searched}")
        print(f"Search time: {duration:.8f} seconds")
        print("----------------------")


if __name__ == "__main__":
    main()

File: artifacts/CS680/CS680_mod8_ENHANCED.rs
use std::collections::HashMap;
use std::fs::File;
use std::io::{self, BufRead, Write};
use std::path::Path;
use std::time::Instant;

struct SearchEngine {
    lookup_table: HashMap<String, i32>,
}

impl SearchEngine {
    fn new() -> Self {
        Self {
            lookup_table: HashMap::new(),
        }
    }

    fn add_record(&mut self, data: String, sequence: i32) {
        self.lookup_table.entry(data).or_insert(sequence);
    }

    fn search(&self, word: &str) -> (Option<i32>, u64, std::time::Duration) {
        let start = Instant::now();
        let result = self.lookup_table.get(word).copied();
        let duration = start.elapsed();
        (result, 1, duration)
    }

    fn len(&self) -> usize {
        self.lookup_table.len()
    }
}

fn load_data(file_path: &str) -> io::Result<SearchEngine> {
    let path = Path::new(file_path);
    let file = File::open(path)?;
    let reader = io::BufReader::new(file);
    let mut engine = SearchEngine::new();

    let mut lines = reader.lines();
    // Skip header
    lines.next();

    for line in lines {
        let line = line?;
        let parts: Vec<&str> = line.split('|').collect();
        if parts.len() == 2 {
            if let Ok(seq) = parts[0].parse::<i32>() {
                engine.add_record(parts[1].to_string(), seq);
            }
        }
    }

    Ok(engine)
}

fn main() {
    println!("--- CS680 Module Eight Search (RUST ENHANCED) ---");
    println!("Architectural Optimization: HashMap (O(1) Search)");

    print!("Enter the numbered data set you want to search (1 - 5): ");
    io::stdout().flush().unwrap();

    let mut dataset_num = String::new();
    io::stdin().read_line(&mut dataset_num).unwrap();
    let dataset_num = dataset_num.trim();

    let file_path = format!(
        "Module Eight Activity Data Sets/Module 8 Data Set-{}.csv",
        dataset_num
    );

    println!(
        "Loading data from {} into an optimized HashMap...",
        file_path
    );
    let start_load = Instant::now();
    let engine = match load_data(&file_path) {
        Ok(e) => e,
        Err(err) => {
            println!("Error loading file: {}", err);
            return;
        }
    };
    let load_duration = start_load.elapsed();

    println!("Data loaded successfully in {:.4?} seconds.", load_duration);
    println!("Total unique records indexed: {}", engine.len());

    loop {
        print!("\nEnter a word to search for (or 'quit' to exit): ");
        io::stdout().flush().unwrap();

        let mut search_word = String::new();
        io::stdin().read_line(&mut search_word).unwrap();
        let search_word = search_word.trim();

        if search_word.to_lowercase() == "quit" {
            println!("Exiting search application.");
            break;
        }

        if search_word.is_empty() {
            continue;
        }

        let (found_seq, records_searched, duration) = engine.search(search_word);

        println!("\n--- Search Results ---");
        if found_seq.is_some() {
            println!("Data: {}", search_word);
        } else {
            println!(
                "The value '{}' was not located within the data set.",
                search_word
            );
        }

        println!("Records searched: {}", records_searched);
        println!("Search time: {:.8?}", duration);
        println!("----------------------");
    }
}
